src/code.c

```diff
@@ -136,11 +136,17 @@ static Stat * ADDR_STAT(Stat stat)
 
 void WRITE_EXPR(Expr expr, UInt idx, UInt val)
 {
+    GAP_ASSERT(expr / sizeof(Expr) + idx <
+               SIZE_BAG(BODY_FUNC(CURR_FUNC())) / sizeof(Expr));
+
```

Comment:
As I wrote in another comment (which is not visible here, which is why I am repeating it here):

Honestly, I don't see how it lines up more closely with how that code works. I'd say to the contrary that what I wrote corresponds more closely to how it works. My version also will catch the case where due to a bug, `expr` is not a multiple of `sizeof(Expr)` (e.g. because somebody passed in an immediate int expression), which one doesn't see if one divides `expr` by `sizeof(Expr)`. 

So, I'd still suggest this instead:
```C
    GAP_ASSERT(expr + idx * sizeof(Expr) < SIZE_BAG(BODY_FUNC(CURR_FUNC())));
```
and similar below.

Reply:
(repeating myelf, so sorry if you see this twice):

I tried to line up with what the code does, and ADDR_EXPR does: `(Expr *)STATE(PtrBody) + expr / sizeof(Expr);`, and doesn't require expr % sizeof(Expr) is 0 (although we could add that as a requirement too).

Reply:
Ahhh, now I understand!

Well, in any case, this is better than no check, so, whatever :-)